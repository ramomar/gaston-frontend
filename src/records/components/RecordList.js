import React from 'react';
import PropTypes from 'prop-types';
import Shapes from '../shapes';
import { Box, Anchor, Text } from 'grommet';
import RecordListItem from './RecordListItem';
import RecordListDaySeparator from './RecordListDaySeparator';
import * as R from 'ramda';
import computeTotalAmount from '../../foundation/computeTotalAmount';

function makeRecordListDateSeparator(day, amountForDay) {
  return (
    <RecordListDaySeparator
      key={day.toLocaleString()}
      day={day}
      amountForDay={amountForDay} />
  );
}

function makeRecordListItem(toRecordReviewScreen) {
  return (record) => {
    return (
      <RecordListItem
        key={record.id}
        record={record}
        toRecordReviewScreen={() => toRecordReviewScreen(record)} />
    );
  };
}

function makeItemsFromRecordGroups(recordGroups, toRecordReviewScreen) {
  const makeItems = ({ day, records }) => R.flatten([
    makeRecordListDateSeparator(day, computeTotalAmount(records)),
    records.map(makeRecordListItem(toRecordReviewScreen))
  ]);

  return R.chain(makeItems, recordGroups);
}

function RecordList(props) {
  return (
    <Box fill='vertical'>
      {props.recordGroups.length === 0 &&
        <Box fill='vertical' justify='center'>
          <Text textAlign='center'>
            Aqui apareceran tus gastos.
          </Text>
        </Box>
      }
      <Box overflow='scroll' pad={{ horizontal: 'medium' }}>
        {makeItemsFromRecordGroups(props.recordGroups, props.toRecordReviewScreen)}
        {!props.isFetching && props.hasMore &&
          <Anchor
            primary
            alignSelf='center'
            margin={{ top: 'small', bottom: 'medium' }}
            style={{ textDecoration: 'underline' }}
            onClick={() => props.moreRecords()}>
            Ver m√°s gastos
          </Anchor>
        }
      </Box>
    </Box>
  );
}

RecordList.propTypes = {
  recordGroups: PropTypes.arrayOf(Shapes.recordGroup).isRequired,
  isFetching: PropTypes.bool.isRequired,
  hasMore: PropTypes.bool.isRequired,
  moreRecords: PropTypes.func.isRequired,
  toRecordReviewScreen: PropTypes.func.isRequired
};

export default RecordList;
