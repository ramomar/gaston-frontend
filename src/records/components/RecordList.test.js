import React from 'react';
import renderer from 'react-test-renderer';
import RecordList from './RecordList';
import { DateTime } from 'luxon';

it('renders correctly', () => {
  const record = {
    id: '0007182d-54cb-42b7-88fc-bbaba51db198',
    amount: 150,
    date: DateTime.fromISO('2021-02-17T12:38:51.700Z'),
    note: 'Cena',
    raw: '{"source": "CHARGE_EMAIL", "type": "EXPENSE", "amount": "150.00", "operation_date": "17/Feb/2021 12:38:51 HORAS", "application_date": "17/Feb/2021", "receiver": null, "channel": {"type": "TPV(COMPRA COMERCIO)"}}'
  };

  const recordGroups = [
    {
      day: DateTime.fromISO('2021-02-17T12:38:51.700Z').startOf('day'),
      records: [record]
    }
  ];

  const component =
    <RecordList
      recordGroups={recordGroups}
      isFetching={false}
      hasMore={true}
      moreRecords={() => null}
      toRecordReviewScreen={() => null} />;

  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});
