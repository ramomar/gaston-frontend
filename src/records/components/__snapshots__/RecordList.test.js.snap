// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 tkzRw"
>
  <div
    className="StyledBox-sc-13pk1d4-0 etTCZD"
  >
    <div
      className="StyledBox-sc-13pk1d4-0 itbAYy"
      style={
        Object {
          "position": "sticky",
          "top": 0,
        }
      }
    >
      <span
        className="StyledText-sc-1sadyjn-0 bnqisW"
      >
        17 de February
      </span>
      <div
        className="StyledBox-sc-13pk1d4-0 cgqXer"
      >
        <span
          className="StyledText-sc-1sadyjn-0 hLzlgx"
          size="small"
        >
          -$
          150.00
        </span>
      </div>
    </div>
    <div
      className="StyledBox-sc-13pk1d4-0 iLrjLl"
      onClick={[Function]}
    >
      <div
        className="StyledBox-sc-13pk1d4-0 dtolEv"
      >
        <span
          className="StyledText-sc-1sadyjn-0 jOAXsc"
        >
          Cena
        </span>
      </div>
      <div
        className="StyledBox-sc-13pk1d4-0 bqDjXT"
      >
        <span
          className="StyledText-sc-1sadyjn-0 isVaoK"
          size="small"
        >
          -$
          150.00
        </span>
        <span
          className="StyledText-sc-1sadyjn-0 gNZQxf"
          size="small"
        >
          06:38
        </span>
      </div>
    </div>
    <a
      className="StyledAnchor-sc-1rp7lwl-0 gxmhec"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      style={
        Object {
          "textDecoration": "underline",
        }
      }
    >
      Ver m√°s gastos
    </a>
  </div>
</div>
`;
